### PR TITLE
FDO All-In-One Server

### PR DESCRIPTION
**Pull request by rftemple**
Adds configurable all in one with docker container
Adds multi-round support

Signed-off-by: Templeton, Randall <randall.f.templeton@intel.com>

### CODE REVIEW

File: component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioFileUploadServlet.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+    final AsyncContext asyncCtx = req.startAsync();
+    asyncCtx.setTimeout(0);
+    new Thread(() -> putAsync(asyncCtx)).start();
+  }
+
+  @Override
+  protected void doDelete(HttpServletRequest req, HttpServletResponse resp) {
+    final int pos = req.getRequestURI().lastIndexOf('/');
+    final String fileName = req.getRequestURI().substring(pos + 1);
```
Comment (DukeDavis12) [inline component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioFileUploadServlet.java:41]: Add checks for directory traversal attack.
We got similar comments from PSEC team during the SDO AIO code review.

File: component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioFileUploadServlet.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    resp.setStatus(HTTP_OK);
+  }
+
+  private void putAsync(AsyncContext asyncCtx) {
+
+    final HttpServletRequest req = (HttpServletRequest) asyncCtx.getRequest();
+    final HttpServletResponse resp = (HttpServletResponse) asyncCtx.getResponse();
+
+    final int pos = req.getRequestURI().lastIndexOf('/');
+    final String fileName = req.getRequestURI().substring(pos + 1);
```
Comment (DukeDavis12) [inline component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioFileUploadServlet.java:60]: Add checks for directory traversal attack.
We got similar comments from PSEC team during the SDO AIO code review.

File: component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioServerApp.java [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+import org.fidoalliance.fdo.protocol.Const;
+import org.h2.server.web.DbStarter;
+import org.h2.server.web.WebServlet;
+
+public class AioServerApp {
+
+  private static final int AIO_PORT = null != AioConfigLoader
+      .loadConfig(AioAppSettings.AIO_PORT)
+      ? Integer.parseInt(AioConfigLoader.loadConfig(AioAppSettings.AIO_PORT))
+      : 8042;
```
Comment (DukeDavis12) [inline component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioServerApp.java:33]: Shouldn't it be 8080 by default?

File: component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioServerApp.java [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+
+    ctx.addParameter(AioAppSettings.TO0_SCHEDULING_ENABLED,
+        AioConfigLoader.loadConfig(AioAppSettings.TO0_SCHEDULING_ENABLED));
+    ctx.addParameter(AioAppSettings.TO0_SCHEDULING_INTREVAL,
+        AioConfigLoader.loadConfig(AioAppSettings.TO0_SCHEDULING_INTREVAL));
+    ctx.addParameter(AioAppSettings.TO0_MAX_WS,
+        AioConfigLoader.loadConfig(AioAppSettings.TO0_MAX_WS));
+
+
+
```
Comment (DukeDavis12) [inline component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioServerApp.java:138]: Empty lines

File: component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioServerApp.java [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+    wrapper = tomcat.addServlet(ctx, "SystemResource", new OwnerSystemResourceServlet());
+    wrapper.setAsyncSupported(true);
+    wrapper.addMapping("/api/v1/device/svi");
+
+    wrapper = tomcat.addServlet(ctx, "sviSettingsServlet",
+        new OwnerSviSettingsServlet());
+    wrapper.addMapping("/api/v1/owner/svi/settings/*");
+
+    wrapper = tomcat.addServlet(ctx, "DiApi", new DiApiServlet());
+    wrapper.addMapping("/api/v1/vouchers/*");
```
Comment (DukeDavis12) [inline component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioServerApp.java:188]: Some API mappings are missing
Eg: /api/v1/assign/?id= ; POST /api/v1/customers/?id=<customer_id>&name=<customer_name>

File: component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioServerApp.java [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+    //setup digest auth
+    LoginConfig config = new LoginConfig();
+    config.setAuthMethod(AioAppSettings.AUTH_METHOD);
+    ctx.setLoginConfig(config);
+    ctx.addSecurityRole(AioAppSettings.AUTH_ROLE);
+    SecurityConstraint constraint = new SecurityConstraint();
+    constraint.addAuthRole(AioAppSettings.AUTH_ROLE);
+    SecurityCollection collection = new SecurityCollection();
+    collection.addPattern("/api/v1/*");
+    constraint.addCollection(collection);
```
Comment (DukeDavis12) [inline component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioServerApp.java:233]: collection.addPattern("/downloads/*") missing

File: component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioContextListener.java [hunk 7]
```diff
@@ -0,0 +1,466 @@
...
+        sc.log("dispatching: " + request.toString());
+      }
+
+      @Override
+      protected void failed(Exception e) {
+        StringWriter writer = new StringWriter();
+        try (PrintWriter pw = new PrintWriter(writer)) {
+          writer.write(e.getMessage());
+          writer.write("\n");
+          e.printStackTrace(pw);
```
Comment (DukeDavis12) [inline component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioContextListener.java:220]: Remove printStackTrace

File: component-samples/demo/aio/aio.env [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+#onDieChace directory must be absolute
+ondie_cache=file:///home/aio/onDieCache/
+ondie_autoupdate=false
+ondie_zip_artifact=
+ondie_check_revocations=true;
+
+aio_to0_max_ws = 28800
+aio_to0_scheduling_enabled=true
+aio_to0_scheduling_interval=300
```
Comment (DukeDavis12) [inline component-samples/demo/aio/aio.env:37]: In AIO, can the to0 scheduling interval be 20 or 30 seconds? We are always changing the to0 scheduling interval value to 20 or 30.

File: component-samples/demo/aio/aio.env [hunk 9]
```diff
@@ -0,0 +1,54 @@
+aio_port=8080
+aio_database_connection_url=jdbc:h2:tcp://localhost:8081/./aio
+aio_database_username=sa
+aio_database_password=
+aio_database_port=8081
+aio_database_driver=org.h2.Driver
+aio_database_allow_others=true
```
Comment (trbehera) [inline component-samples/demo/aio/aio.env:7]: The default behavior should be false.

We should also print a warning message if this is enabled, stating that "This property should NOT be set to true in production deployments."

File: component-samples/demo/aio/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+RUN mkdir -p /home/aio/lib
+WORKDIR /home/aio/
+
+# Setup aio dependencies
+RUN mkdir -p /home/aio/onDieCache
+COPY ./lib ./lib/
+COPY ./certs ./certs
+COPY aio.jar .
+COPY manufacturer_keystore.p12 .
```
Comment (DukeDavis12) [inline component-samples/demo/aio/Dockerfile:23]: COPY ./resource ./resource

File: component-samples/demo/aio/aio.env [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+manufacturer_keystore_type=PKCS12
+manufacturer_keystore_password=MfgKs@3er
+
+#onDieChace directory must be absolute
+ondie_cache=file:///home/aio/onDieCache/
+ondie_autoupdate=false
+ondie_zip_artifact=
+ondie_check_revocations=true;
+
+aio_to0_max_ws = 28800
```
Comment (DukeDavis12) [inline component-samples/demo/aio/aio.env:35]: should be `aio_to0_max_ws=28800`

File: component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioConfigLoader.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+import org.apache.commons.configuration2.FileBasedConfiguration;
+import org.apache.commons.configuration2.PropertiesConfiguration;
+import org.apache.commons.configuration2.SystemConfiguration;
+import org.apache.commons.configuration2.builder.FileBasedConfigurationBuilder;
+import org.apache.commons.configuration2.builder.fluent.Parameters;
+import org.apache.commons.configuration2.ex.ConfigurationException;
+
+public class AioConfigLoader {
+
+  private static final String configurationFile = "application.properties";
```
Comment (trbehera) [inline component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioConfigLoader.java:13]: Other components use .env files for loading the configurations. We should change this to follow that pattern.

File: component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioContextListener.java [hunk 13]
```diff
@@ -0,0 +1,466 @@
...
+          key.get(), to01Payload.toBytes(), cs.getCoseAlgorithm(ownerPublicKey));
+    } catch (IOException e) {
+      throw new DispatchException(e);
+    }
+
+    return singedBlob.toBytes();
+  }
+
+
+  private void scheduleTo0(ServletContext sc, DataSource ds, UUID guid, OwnerDbTo0Util to0Util) {
```
Comment (trbehera) [inline component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioContextListener.java:453]: If this is unused, it can be removed.

File: component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioDbManager.java [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+  public void createTables(DataSource ds) {
+    try (Connection conn = ds.getConnection();
+        Statement stmt = conn.createStatement()) {
+
+      String sql =
+          "CREATE TABLE IF NOT EXISTS "
+              + "TO2_CONFIG ("
+              + "ID INT NOT NULL DEFAULT 1, "
+              + "INITIALIZED BOOLEAN NOT NULL DEFAULT FALSE, "
+              + "RV_BLOB VARCHAR(4096) NOT NULL DEFAULT 'http://localhost:8080', "
```
Comment (trbehera) [inline component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioDbManager.java:35]: This should get updated to CBOR RVInfo Blob.

File: component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioDbManager.java [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+        + "VALUES (?,?,?,?,?,?,?);";
+
+    try (Connection conn = ds.getConnection();
+        PreparedStatement pstmt = conn.prepareStatement(sql)) {
+
+      pstmt.setString(1, guid);
+      pstmt.setBytes(2, signedBlob);
+      pstmt.setString(3, fingerPrint);
+      pstmt.setBytes(4, deviceX509);
+      pstmt.setInt(5, 315400000);
```
Comment (trbehera) [inline component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioDbManager.java:99]: It would be good to define 315400000 as a constant (explaining how we reach this number) and use here.

File: component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioDispatchListener.java [hunk 16]
```diff
@@ -0,0 +1,11 @@
+package org.fidoalliance.fdo.sample;
```
Comment (trbehera) [inline component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioDispatchListener.java:1]: We would need to add Apache v2 license header and copyright in all new files.

File: component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioDbManager.java [hunk 17]
```diff
@@ -0,0 +1,193 @@
+package org.fidoalliance.fdo.sample;
```
Comment (DukeDavis12) [inline component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioDbManager.java:1]: Licence header not added.

File: component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioFileDownloadServlet.java [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Allows download of files using curl or wget.
+ */
+public class AioFileDownloadServlet extends HttpServlet {
+
+  // common http setting
+  public static final int HTTP_OK = 200;
```
Comment (trbehera) [inline component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioFileDownloadServlet.java:19]: Use HttpServletResponse.SC_xxx codes instead.

File: component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioContextListener.java [hunk 19]
```diff
@@ -0,0 +1,466 @@
+package org.fidoalliance.fdo.sample;
```
Comment (DukeDavis12) [inline component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioContextListener.java:1]: licence header not added.

File: component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioServerApp.java [hunk 20]
```diff
@@ -0,0 +1,279 @@
...
+        AioConfigLoader.loadConfig(AioAppSettings.DB_NEW_DEVICE_SQL));
+
+    // hard-coded H2 config
+    // To enable remote connections to the DB set
+    // db.tcpServer=-tcp -tcpAllowOthers -ifNotExists -tcpPort
+    // This creates a security hole in the system.
+    // Not recommended to use especially on production system
+    ctx.addParameter("db.tcpServer", "-tcp -ifNotExists -tcpPort "
+        + AioConfigLoader.loadConfig(AioAppSettings.DB_PORT));
+
```
Comment (trbehera) [inline component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioServerApp.java:85]: Empty lines .. can be removed.

File: component-samples/demo/aio/Dockerfile [hunk 21]
```diff
@@ -0,0 +1,32 @@
+# Copyright 2020 Intel Corporation
+# SPDX-License-Identifier: Apache 2.0
+
+FROM ubuntu:20.04
+ARG _JAVA_OPTIONS
+ENV _JAVA_OPTIONS=${_JAVA_OPTIONS}
+RUN apt-get update && apt-get install -y openjdk-11-jdk libsofthsm2 softhsm softhsm-common opensc wget haveged
```
Comment (trbehera) [inline component-samples/demo/aio/Dockerfile:7]: Should include `apt-get -y upgrade` to ensure we have latest packages installed in the docker container.

File: component-samples/demo/aio/Dockerfile [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+WORKDIR /home/aio/
+
+# Setup aio dependencies
+RUN mkdir -p /home/aio/onDieCache
+COPY ./lib ./lib/
+COPY ./certs ./certs
+COPY aio.jar .
+COPY manufacturer_keystore.p12 .
+COPY owner_keystore.p12 .
+COPY owner_customer1.pem .
```
Comment (trbehera) [inline component-samples/demo/aio/Dockerfile:25]: owner_customer1.pem is missing the commit.

File: component-samples/demo/aio/docker-compose.yml [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+    build:
+      context: .
+      dockerfile: Dockerfile
+      args:
+        - _JAVA_OPTIONS
+    restart: on-failure:5
+    env_file:
+      - ./aio.env
+    #volumes:
+     # - ./target:/home/aio/target
```
Comment (trbehera) [inline component-samples/demo/aio/docker-compose.yml:24]: In unused, this can be removed.

File: component-samples/demo/aio/init.sql [hunk 24]
```diff
@@ -0,0 +1,30 @@
...
+
+INSERT INTO SYSTEM_MODULE_RESOURCE (CONTENT,CONTENT_TYPE_TAG,PRIORITY ) VALUES('F5','fdo_sys:active',0  );
+//INSERT INTO SYSTEM_MODULE_RESOURCE (CONTENT,GUID_TAG,FILE_NAME_TAG,CONTENT_TYPE_TAG,PRIORITY ) VALUES('68656c6c6f2066646f20776f726c64','2e1418eb-c810-49c4-a0db-91c4d0a58306','test.txt','fdo_sys:filedesc',1  )
+//INSERT INTO SYSTEM_MODULE_RESOURCE (CONTENT,GUID_TAG,CONTENT_TYPE_TAG,PRIORITY ) VALUES('8363636D64622F636B737461727475702E626174','2e1418eb-c810-49c4-a0db-91c4d0a58306','fdo_sys:exec',2  )
+          
+//INSERT INTO TO2_SETTINGS  (ID,DEVICE_SERVICE_INFO_MTU_SIZE, OWNER_MTU_THRESHOLD  ) VALUES(1,1300,10400  );
+
+
+//the sets the to2 owner blob returned during to1
+UPDATE TO2_CONFIG SET RV_BLOB = 'http://localhost:8080?ipaddress=127.0.0.1';
```
Comment (trbehera) [inline component-samples/demo/aio/init.sql:27]: Replaced with CBOR RVInfo blob.

File: component-samples/demo/aio/init.sql [hunk 25]
```diff
@@ -0,0 +1,30 @@
...
+//set owner to2 replacement keys
+INSERT INTO OWNER_CUSTOMERS (CUSTOMER_ID,NAME ,KEYS) VALUES (1,'owner',FILE_READ('./owner_customer1.pem',NULL) );
+
+
+
+INSERT INTO SYSTEM_MODULE_RESOURCE (CONTENT,CONTENT_TYPE_TAG,PRIORITY ) VALUES('F5','fdo_sys:active',0  );
+//INSERT INTO SYSTEM_MODULE_RESOURCE (CONTENT,GUID_TAG,FILE_NAME_TAG,CONTENT_TYPE_TAG,PRIORITY ) VALUES('68656c6c6f2066646f20776f726c64','2e1418eb-c810-49c4-a0db-91c4d0a58306','test.txt','fdo_sys:filedesc',1  )
+//INSERT INTO SYSTEM_MODULE_RESOURCE (CONTENT,GUID_TAG,CONTENT_TYPE_TAG,PRIORITY ) VALUES('8363636D64622F636B737461727475702E626174','2e1418eb-c810-49c4-a0db-91c4d0a58306','fdo_sys:exec',2  )
+          
+//INSERT INTO TO2_SETTINGS  (ID,DEVICE_SERVICE_INFO_MTU_SIZE, OWNER_MTU_THRESHOLD  ) VALUES(1,1300,10400  );
```
Comment (trbehera) [inline component-samples/demo/aio/init.sql:23]: 10400 => 8192

File: component-samples/demo/aio/java.security [hunk 26]
```diff
@@ -0,0 +1,1209 @@
+#
```
Comment (trbehera) [inline component-samples/demo/aio/java.security:1]: Considering we are not using soft HSM, this file is not used anymore and can be removed.

File: component-samples/demo/aio/pkcs11.cfg [hunk 27]
```diff
@@ -0,0 +1,3 @@
+name = softhsm
```
Comment (trbehera) [inline component-samples/demo/aio/pkcs11.cfg:1]: Not used, can be removed.

File: component-samples/demo/aio/resources/hello-world.sh [hunk 28]
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
+#
+# Copyright 2020 Intel Corporation
+# SPDX-License-Identifier: Apache 2.0
+
+echo "Fido Device Onboarding completed sucessfully!" > out.exe
```
Comment (trbehera) [inline component-samples/demo/aio/resources/hello-world.sh:6]: out.exe => out.txt

File: protocol-samples/http-client-to0-sample/src/main/java/org/fidoalliance/fdo/sample/To0ClientApp.java [hunk 29]
```diff
@@ -31,29 +31,29 @@
   private static final String RV_BLOB = "http://localhost:8042?ipaddress=127.0.0.1";
```
Comment (trbehera) [inline protocol-samples/http-client-to0-sample/src/main/java/org/fidoalliance/fdo/sample/To0ClientApp.java:31]: Looks like unrelated change, can be removed.

File: protocol-samples/http-client-to1-sample/src/main/java/org/fidoalliance/fdo/sample/To1ClientApp.java [hunk 30]
```diff
@@ -39,10 +39,10 @@
       + "-----END EC PRIVATE KEY-----";
 
   private static final String deviceCreds = ""
-      + "87f5186458401124dc2801de1a8d31978c0efc205871469cbe851eb226838219ffa66f0c1ffc3550bdfdb50fe"
```
Comment (trbehera) [inline protocol-samples/http-client-to1-sample/src/main/java/org/fidoalliance/fdo/sample/To1ClientApp.java:42]: Unrelated changes, can be ignored.

File: protocol-samples/http-client-to2-sample/src/main/java/org/fidoalliance/fdo/sample/To2ClientApp.java [hunk 31]
```diff
@@ -50,17 +50,16 @@
       + "-----END EC PRIVATE KEY-----";
 
   private static final String deviceCreds = ""
-      + "87f5186458401124dc2801de1a8d31978c0efc205871469cbe851eb226838219ffa66f0c1ffc3550bdfdb50fe"
```
Comment (trbehera) [inline protocol-samples/http-client-to2-sample/src/main/java/org/fidoalliance/fdo/sample/To2ClientApp.java:53]: Unrelated change, can be removed.

File: util/storage-samples/storage-owner-sample/src/main/java/org/fidoalliance/fdo/storage/OwnerSysModule.java [hunk 32]
```diff
@@ -65,11 +70,20 @@ public Composite getState() {
 
   @Override
   public void setServiceInfo(Composite kvPair, boolean isMore) {
+    String key = kvPair.getAsString(Const.FIRST_KEY);
+    Object value = kvPair.get(Const.SECOND_KEY);
 
+    switch (key) {
```
Comment (trbehera) [inline util/storage-samples/storage-owner-sample/src/main/java/org/fidoalliance/fdo/storage/OwnerSysModule.java:76]: This switch case is not performing any operation.

File: component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioServerApp.java (hunk 5 above)
Comment (rftemple) [reply component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioServerApp.java:188]: all api's  should be covered by /api/v1/*    all servlet api patterns are "/api/v1/*" e.g. /api/v1/assign/*""

File: component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioServerApp.java [hunk 33]
```diff
@@ -0,0 +1,279 @@
...
+    wrapper.addMapping("/api/v1/vouchers/*");
+    wrapper.setAsyncSupported(true);
+
+    wrapper = tomcat.addServlet(ctx, "AssignCustomerApi", new AssignCustomerServlet());
+    wrapper.addMapping("/api/v1/assign/*");
+    wrapper.setAsyncSupported(true);
+
+    wrapper = tomcat.addServlet(ctx, "UpdateRvInfoApi", new RvInfoServlet());
+    wrapper.addMapping("/api/v1/rvinfo/*");
+    wrapper.setAsyncSupported(true);
```
[the remark this answers is not in the document]
Comment (rftemple) [reply component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioServerApp.java:197]: all api's  should be covered by /api/v1/*    all servlet api patterns are "/api/v1/*" e.g. /api/v1/rvinfo/*

File: component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioServerApp.java (hunk 6 above)
Comment (rftemple) [reply component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioServerApp.java:233]: "/downloads/*" is intentionally missing because its the only API that  does not require any authentication.  Its for downloading agents/scripts that have no secrets in them.

all api's  should be covered by /api/v1/*    all servlet api patterns are "/api/v1/*" e.g. /api/v1/device/svi""

File: component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioConfigLoader.java (hunk 12 above)
Comment (rftemple) [reply component-samples/aio/src/main/java/org/fidoalliance/fdo/sample/AioConfigLoader.java:13]: aio uses .env same as other components.  It will use application.properties only if present - all other component have this same definition

File: util/storage-samples/storage-owner-sample/src/main/java/org/fidoalliance/fdo/storage/OwnerDbManager.java [hunk 34]
```diff
@@ -557,7 +557,9 @@ public Composite getSystemResources(DataSource ds,
         + "(ARCHITECTURE_TAG = ? OR ARCHITECTURE_TAG IS NULL ) AND "
         + "(CONTENT_TYPE_TAG = 'fdo_sys:filedesc' "
         + "OR CONTENT_TYPE_TAG = 'fdo_sys:exec' "
-        + "OR CONTENT_TYPE_TAG = 'fdo_sys:active') "
+        + "OR CONTENT_TYPE_TAG = 'fdo_sys:active' "
```
Comment (trbehera) [inline util/storage-samples/storage-owner-sample/src/main/java/org/fidoalliance/fdo/storage/OwnerDbManager.java:560]: Document additional fdo_sys keys

2 more keys have been added to fdo_sys module (ismore and isdone). These need to be added to github.io documentation for fdo_sys module.

File: util/storage-samples/storage-owner-sample/src/main/java/org/fidoalliance/fdo/storage/OwnerDevMod.java [hunk 35]
```diff
@@ -110,6 +110,8 @@ public void setServiceInfo(Composite kvPair, boolean isMore) {
       covertModuleList();
       new OwnerDbManager().addDeviceInfo(dataSource, state);
     }
+
```
Comment (trbehera) [inline util/storage-samples/storage-owner-sample/src/main/java/org/fidoalliance/fdo/storage/OwnerDevMod.java:113]: Unrelated change, can be removed.

**Review by DukeDavis12 — APPROVED** [review]

File: util/storage-samples/storage-owner-sample/src/main/java/org/fidoalliance/fdo/storage/OwnerSysModule.java [hunk 36]
```diff
@@ -61,11 +67,31 @@ public Composite getState() {
 
   @Override
   public void setServiceInfo(Composite kvPair, boolean isMore) {
-
+    String key = kvPair.getAsString(Const.FIRST_KEY);
+    Object value = kvPair.get(Const.SECOND_KEY);
+
+    switch (key) {
+      case FdoSys.KEY_ACTIVE:
```
Comment (prateekcr) [inline util/storage-samples/storage-owner-sample/src/main/java/org/fidoalliance/fdo/storage/OwnerSysModule.java:74]: If active=false, then:
1. Should state's 4th key be set to false explicitly, along with the 5th (false) and 6th keys (true), since there's nothing to be sent?
2. Since the 2nd key is set to 0 here, the conditions at methods 'isMore()' and 'nextMessage()', that is condition 'if (resIndex < 0)' won't be triggered. Should the 2nd key be set to -1, or the condition be changed to 'if (resIndex <= 0)' ?

Comment (rftemple) [reply util/storage-samples/storage-owner-sample/src/main/java/org/fidoalliance/fdo/storage/OwnerSysModule.java:74]: yes fixing point 1)  but regarding point 2) resIndex <0 is a special condition that indicates that the module is not ready to respond (e.g the devmod as not finished).  resIndex >=0 means the module is ready to respond but may or may not be done.   IsMore() always returns done=true and ismore=false if resIndex >= resList.Length.  else if (resIndex < IndexLength we still have messages to process.   I did add a resList.clear() when active=false.  This will always trigger checkMessage to return done=true and isMore=false for a future requests.

**Review by KiranSukhavasi — DISMISSED** [review]

**Review by DukeDavis12 — APPROVED** [review]
